framework/web/View.php

```diff
@@ -516,6 +516,45 @@ public function registerJsFile($url, $options = [], $key = null)
         }
     }
 
+    /**
+     * Registers a JS code block defining a variable. The name of variable will be
+     * used as key, preventing duplicated variable names.
+     *
+     * @param string $name Name of the variable
+     * @param array|string $value Value of the variable
+     * @param int $position the position in a page at which the JavaScript variable should be inserted.
+     * The possible values are:
+     *
+     * - [[POS_HEAD]]: in the head section. This is the default value.
+     * - [[POS_BEGIN]]: at the beginning of the body section.
+     * - [[POS_END]]: at the end of the body section.
+     * - [[POS_LOAD]]: enclosed within jQuery(window).load().
+     *   Note that by using this position, the method will automatically register the jQuery js file.
+     * - [[POS_READY]]: enclosed within jQuery(document).ready().
+     *   Note that by using this position, the method will automatically register the jQuery js file.
+     *
+     * @param string $key the key that identifies the JS code block. If null, it will use
+     * $js as the key. If two JS code blocks are registered with the same key, the latter
+     * will overwrite the former.
+     *
+     * @throws InvalidConfigException If variable name does not match ECMA requirements.
+     * @since 2.0.14
+     */
+    public function registerJsVar($name, $value, $position = self::POS_HEAD)
+    {
+        $validator = new \yii\validators\RegularExpressionValidator([
+            'pattern' => '/^[^a-zA-Z_]+|[^a-zA-Z_0-9]+/',
```

Comment:
This regexp is incorrect. See https://mathiasbynens.be/notes/javascript-identifiers#examples

I'm not sure if we should validate it at all. It should be a developer responsibility to use valid name for variable.

Reply:
Reminds me of https://stackoverflow.com/questions/48270127/can-a-1-a-2-a-3-ever-evaluate-to-true :)

I'd say it's good to do this kind of validation, even if JavaScript could use other chars in the variable name - it's just crazy and from my understanding it will be in ES6 that this is no longer allowed. See also these numerous issues with browsers.

Reply:
Right now this validation:
* allows invalid variable names (like keywords),
* disallows valid names (not only insane `var ಠ_ಠ = eval;` but also pretty common `var $ = jQuery;`.

It simply doesn't work and create unnecessary overhead.

> I'd say it's good to do this kind of validation, even if JavaScript could use other chars in the variable name - it's just crazy.

I'm not sure if Yii as PHP framework should force some name conventions for JavaScript variables.

Reply:
Here's a regexp for the validation: https://stackoverflow.com/a/9337047/291573 - it's 11kB - 🤣 

When I think about compatibility to PHP or among browsers and various JS-engines, I'd rather add a strict validation - but I do not really care. If someone thinks it's a good idea to name variables in non-7-bit-ASCII, so be it...

Reply:
Addon: https://mathiasbynens.be/notes/javascript-identifiers-es6

Some headlines from examples from that page:

// Valid in ES5, but invalid in an ES6 module context:
// Valid in ES5 and ES6:
// Invalid in ES5, but valid in ES6:
// Valid in ES5 & Unicode v5.1.0+, but invalid in ES6:
// Invalid in ES5 and ES6:
// Valid in ES5, but only works in some ES5 engines (i.e. those with Unicode data from v3.2.0 or more recent):
// Valid in ES5, but only works in some ES5 engines (i.e. those with Unicode data from v5.1.0 or more recent):
// Invalid in ES5 and ES6, but supported in old ES5 engines:

Reply:
Nice list. We'd better remove validation.